File: wads/src/main/java/com/walhalla/domain/interactors/impl/AdvertInteractorImpl.java
package com.walhalla.domain.interactors.impl;

import android.view.View;
import android.view.ViewGroup;

import com.google.android.gms.ads.AdView;
import com.walhalla.boilerplate.domain.executor.Executor;
import com.walhalla.boilerplate.domain.executor.MainThread;
import com.walhalla.boilerplate.domain.interactors.base.AbstractInteractor;

import com.walhalla.domain.interactors.AdvertInteractor;
import com.walhalla.domain.repository.AdvertRepository;
import com.walhalla.library.AdMobCase;

/**
 * AbstractInteractor
 * запустит все в фоновом потоке...
 */

public class AdvertInteractorImpl extends AbstractInteractor
        implements AdvertInteractor {

    private AdvertRepository mAdvertRepository;

    public AdvertInteractorImpl(
            Executor threadExecutor,
            MainThread mainThread,
            AdvertRepository repository) {
        super(threadExecutor, mainThread);
        mAdvertRepository = repository;
    }

    //Отдаем результат
    public void notifyError() {
        mMainThread.post(() -> {
            //mCallback.onRetrievalFailed("Nothing to welcome you with :(");
        });
    }

    public void postMessage(final String msg) {
        mMainThread.post(() -> {
            //mCallback.onMessageRetrieved(msg);
        });
    }

    @Override
    public void selectView(ViewGroup viewGroup, final Callback<View> callback) {
        Runnable runnable = () -> {
//            try {
//                String name = Thread.currentThread().getName();
//                Log.i(TAG, "Foo " + name);
//                TimeUnit.SECONDS.sleep(1);
//                Log.i(TAG, "Bar " + name);

                final int id = viewGroup.getId();
                final AdView adView = mAdvertRepository.getNewAdsBanner(viewGroup);

                if (adView == null) {
                    callback.onRetrievalFailed("View not found");
                    return;
                }
                mMainThread.post(() -> {
                    callback.onMessageRetrieved(id, adView);
                    AdMobCase.interstitialBannerRequest(adView);
                });
//            } catch (InterruptedException e) {
//                e.printStackTrace();
//            }
        };
        mThreadExecutor.submit(runnable);
    }

    @Override
    public void run() {}
}

File: features/wads/src/main/java/com/walhalla/domain/interactors/impl/AdvertInteractorImpl.kt
package com.walhalla.domain.interactors.impl

import android.view.View
import android.view.ViewGroup
//import com.walhalla.boilerplate.domain.executor.Executor
//import com.walhalla.boilerplate.domain.executor.MainThread
//import com.walhalla.boilerplate.domain.interactors.base.AbstractInteractor
import com.walhalla.domain.interactors.AdvertInteractor
import com.walhalla.domain.repository.AdvertRepository
import com.walhalla.library.AdMobCase.interstitialBannerRequest
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.MainScope
import kotlinx.coroutines.launch

/**
 * AbstractInteractor
 * запустит все в фоновом потоке...
 */
class AdvertInteractorImpl(
    //threadExecutor: Executor, mainThread: MainThread,

    var threadExecutor: CoroutineScope = CoroutineScope(Dispatchers.IO),
    var mainThread: CoroutineScope = MainScope()//CoroutineScope(Dispatchers.Main)

    , var mAdvertRepository: AdvertRepository
) :
/*AbstractInteractor(threadExecutor, mainThread),*/ AdvertInteractor {

//    //Отдаем результат
//    fun notifyError() {
//        mMainThread.post {}
//    }
//
//    fun postMessage(msg: String) {
//        mMainThread.post {}
//    }

    override fun selectView(viewGroup: ViewGroup, callback: AdvertInteractor.Callback<View>) {
        threadExecutor.launch {
            val adView = mAdvertRepository.getNewAdsBanner(viewGroup)

            if (adView == null) {
                mainThread.launch {
                    callback.onRetrievalFailed("View not found")
                }
                return@launch
            }

            mainThread.launch {
                callback.onMessageRetrieved(viewGroup.id, adView)
                interstitialBannerRequest(adView)
            }
        }
    }

    //override fun run() {}
}
